components/concept/engine/src/main/java/mozilla/components/concept/engine/manifest/WebAppManifestParser.kt

```diff
@@ -44,22 +44,34 @@ class WebAppManifestParser {
     /**
      * Parses the provided JSON and returns a [WebAppManifest] (wrapped in [Result.Success] if parsing was successful.
      * Otherwise [Result.Failure].
+     *
+     * Gecko performs some initial parsing on the Web App Manifest, so the [JSONObject] we work with
+     * does not match what was originally provided by the website. Gecko:
+     * - Changes relative URLs to be absolute
+     * - Changes some space-separated strings into arrays (purpose, sizes)
+     * - Changes colors to follow Android format (#AARRGGBB)
+     * - Removes invalid enum values (ie display: halfscreen)
+     * - Ensures display, dir, start_url, and scope always have a value
+     * - Trims most strings (name, short_name, ...)
+     * See See https://searchfox.org/mozilla-central/source/dom/manifest/ManifestProcessor.jsm
      */
     fun parse(json: JSONObject): Result {
         return try {
+            val startUrl = json.getString("start_url")
+
             val shortName = json.tryGetString("short_name")
             val name = json.tryGetString("name") ?: shortName
                 ?: return Result.Failure(JSONException("Missing manifest name"))
 
             Result.Success(WebAppManifest(
                 name = name,
                 shortName = shortName,
-                startUrl = json.getString("start_url"),
+                startUrl = startUrl,
                 display = parseDisplayMode(json),
                 backgroundColor = parseColor(json.tryGetString("background_color")),
                 description = json.tryGetString("description"),
                 icons = parseIcons(json),
-                scope = json.tryGetString("scope"),
+                scope = json.tryGetString("scope") ?: startUrl,
```

Comment:
How does this affect manifest that we already have saved to disk? Could we have saved a `null` at any point in the past?

Using the start url as default would mean that we always switch to a custom tab once the user leaves the start url? Is that the expected default? I wonder if we need to be more protective against changes in Gecko(View)? Since the preprocessing is not part of the standard we can't always rely on it?

Reply:
Yes, the default isn't perfect. I put in a poor default because we should always get a value from storage and Gecko...so I should probably change this to throw if the scope isn't defined :)

It might be worthwhile to schedule a meeting with the Gecko side and decide what to do long term. I think its OK to rely on the preprocessing, especially since it differs the standard in some areas like `sizes`. The standard is being updated to include a concept of a "processed manifest", which is closer to what we're working with here.